lib/playdate/games/sudoku/root_component.ex

```diff
@@ -0,0 +1,205 @@
+defmodule Playdate.Games.Sudoku.RootComponent do
+  @moduledoc """
+  The liveView component for the Sudoku UI.
+  """
+  use Surface.LiveView
+
+  alias Playdate.AppHelpers
+  alias PlaydateWeb.Router.Helpers, as: Routes
+  alias Playdate.Games.Sudoku.{Keyboard, Server}
+  alias Playdate.PubSub
+
+  @impl true
+  @spec mount(map, map, Socket.t()) :: {:ok, Socket.t()}
+  def mount(_params, %{"instance" => instance} = session, socket) do
+    app_user = AppHelpers.app_user_from_session(session)
+    state = Server.get_state(instance)
+
+    other_players =
+      for player <- instance.app_users,
+          player.id !== app_user.id,
+          do: {player.id, player.nickname}
+
+    socket =
+      socket
+      |> assign(state)
+      |> assign(
+        me: app_user.id,
+        nickname: app_user.nickname,
+        other_players: other_players,
+        instance: instance,
+        active_pos: nil,
+        time: ~T[00:00:00]
+      )
+
+    if connected?(socket) do
+      PubSub.subscribe("sudoku:#{instance.id}")
+    end
+
+    {:ok, socket}
+  end
+
+  @doc false
+  @impl true
+  def handle_event("select_pos", %{"pos" => pos}, socket) do
+    position =
+      if String.to_integer(pos) in socket.assigns.editable_pos,
+        do: String.to_integer(pos),
+        else: nil
+
+    socket =
+      socket
+      |> assign(active_pos: position)
+
+    {:noreply, socket}
+  end
+
+  def handle_event("keyboard", %{"key" => key}, socket)
+      when not is_nil(socket.assigns.active_pos) do
+    case key do
+      "undo" ->
+        Server.undo(socket.assigns.instance, socket.assigns.me, socket.assigns.active_pos)
+        {:noreply, socket}
+
+      _ ->
+        guess = String.to_integer(key)
+        Server.play(socket.assigns.instance, socket.assigns.me, socket.assigns.active_pos, guess)
+        {:noreply, socket}
+    end
+  end
+
+  def handle_event("keyboard", %{}, socket), do: {:noreply, socket}
+
+  @doc false
+  @impl true
+  def handle_info({:update, new_state}, socket) do
+    socket =
+      socket
+      |> assign(new_state)
+
+    {:noreply, socket}
+  end
+
+  def handle_info(:exit, socket) do
+    {:noreply, push_redirect(socket, to: Routes.lobby_index_path(socket, :index))}
+  end
+
+  def handle_info(:timer, socket) do
+    if socket.assigns.status == :playing do
+      Process.send_after(self(), :timer, 1_000)
+
+      time =
+        Time.add(socket.assigns.time, 1, :second)
+        |> Time.truncate(:second)
+
+      {:noreply, assign(socket, :time, time)}
+    else
+      {:noreply, socket}
+    end
+  end
+
+  @doc false
+  @impl true
+  def render(assigns) do
```

Comment:
```suggestion
  @spec render(Socket.assigns) :: Phoenix.LiveView.Rendered.t()
  def render(assigns) do
```

(worth aliasing in `Rendered` as well).